### PR TITLE
InputParameters: refactor/consolidate param metadata into single struct

### PR DESCRIPTION
**Pull request by rwcarlsen**
This makes the code easier to understand and maintain and is preparation
for resolving #10357.

### CODE REVIEW

File: framework/include/utils/InputParameters.h [hunk 1]
```diff
@@ -687,24 +679,51 @@ class InputParameters : public Parameters
...
-
-  /// Syntax for command-line parameters
-  std::map<std::string, std::vector<std::string>> _syntax;
-
-  /// The names of the parameters organized into groups
-  std::map<std::string, std::string> _group;
-
-  /// The map of functions used for range checked parameters
-  std::map<std::string, std::string> _range_functions;
+  struct Metadata
```
Comment (permcody) [inline framework/include/utils/InputParameters.h]: Sorry, you will need to add leading underscores to all members here per the MOOSE coding guidelines.

File: framework/include/utils/InputParameters.h [hunk 2]
```diff
@@ -687,24 +679,51 @@ class InputParameters : public Parameters
...
+    /// True for parameters that are required (i.e. will cause an abort if not supplied)
+    bool required = false;
+    /**
+     * Whether the parameter is either explicitly set or provided a default value when added
+     * Note: We do not store MooseEnum names in valid params, instead we ask MooseEnums whether
+     *       they are valid or not.
+     */
+    bool valid = false;
+    /// The set of parameters that will NOT appear in the the dump of the parser tree
+    bool isprivate = false;
```
Comment (permcody) [inline framework/include/utils/InputParameters.h]: `is_private`

**Review by permcody — CHANGES_REQUESTED** [review]
This looks good! You just need to follow the underscore conventions for members in the struct.

File: framework/include/utils/InputParameters.h (hunk 1 above)
Comment (rwcarlsen) [reply framework/include/utils/InputParameters.h]: done.

**Review by permcody — APPROVED** [review]

**Comment by YaqiWang** [comment]
This seems having a noticeable effect on timing of Rattlesnake benchmarks: https://hpcsc.inl.gov/ssl/RATTLESNAKE/timing/, should we be bothered?

**Comment by rwcarlsen** [comment]
That PR should only have any impact on simulation setup time (and I expect
the impact should be small).  Is that benchmark's setup heavy?

**Comment by YaqiWang** [comment]
Yes, saaf_sn test is heavier than the others, its setup summary is
```
      Variables - 40
   AuxVariables - 1
        Kernels - 160
     AuxKernels - 1
            BCs - 160
    UserObjects - 1
   AuxMaterials - 1
```
while neutron_saaf_sn is
```
      Variables - 12
   AuxVariables - 1
        Kernels - 48
     AuxKernels - 1
            BCs - 36
 Postprocessors - 2
    UserObjects - 1
   AuxMaterials - 2
```
It is size is not that big in Rattlesnake though (we could have thousands of variables). I would say something needs possibly to be improved.

**Comment by YaqiWang** [comment]
BTW, the current MR in Rattlesnake by you is failing precheck. Is it related with this?

**Comment by rwcarlsen** [comment]
Rattlesnake PR is suppose to fail until another MOOSE PR gets in - don't
worry about it - I'll ping when it's ready.  Yaqi - could you provide the
setup summary for before the PR too - to see what in the balance may have
changed?

**Comment by YaqiWang** [comment]
I scaled up the saaf_sn test to 
```
      Variables - 480
   AuxVariables - 1
        Kernels - 1920
     AuxKernels - 1
            BCs - 1920
    UserObjects - 1
   AuxMaterials - 1
```
and ran it before (7de6cd863b051d6ac004924c5b612157a7139c11) and after (7de6cd863b051d6ac004924c5b612157a7139c11) this PR, but did not see any noticeable increase in cpu time on my mac with five runs. So I am not sure if it is worthwhile to do any further investigations. I tend to ignore it for now since it is only affecting setup and possibly platform dependent. Thanks guys.

**Pull request merged**